src/client/serialize/daos_serialize.c

```diff
@@ -0,0 +1,950 @@
+		rc = -DER_IO;
+		D_ERROR("failed to write attributes "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+out:
+	for (i = 0; i < len_acl; i++) {
+		D_FREE(acl_strs[i]);
+	}
+	D_FREE(acl_strs);
+
+	/* close hdf5 objects */
+	if (usr_attr > 0)
+		H5Aclose(usr_attr);
+	if (attr_dtype > 0)
+		H5Tclose(attr_dtype);
+	if (attr_dspace > 0)
+		H5Sclose(attr_dspace);
+	return rc;
+}
+
+static int
+serialize_str(hid_t file_id, struct daos_prop_entry *entry,
+	      const char *prop_str)
+{
+
+	int	rc = 0;
+	hid_t	status = 0;
+	hsize_t	attr_dims[1];
+	hid_t	attr_dtype = 0;
+	hid_t	attr_dspace = 0;
+	hid_t	usr_attr = 0;
+
+	if (entry == NULL || entry->dpe_str == NULL) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+
+	attr_dims[0] = 1;
+	attr_dtype = H5Tcopy(H5T_C_S1);
+	if (attr_dtype < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	status = H5Tset_size(attr_dtype, strlen(entry->dpe_str) + 1);
+	if (status < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	status = H5Tset_strpad(attr_dtype, H5T_STR_NULLTERM);
+	if (status < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	attr_dspace = H5Screate_simple(1, attr_dims, NULL);
+	if (attr_dspace < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+
+	usr_attr = H5Acreate2(file_id, prop_str, attr_dtype, attr_dspace,
+			      H5P_DEFAULT, H5P_DEFAULT);
+	if (usr_attr < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	status = H5Awrite(usr_attr, attr_dtype, entry->dpe_str);
+	if (status < 0) {
+		rc = -DER_IO;
+		D_ERROR("failed to write attribute "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+out:
+	/* close hdf5 objects */
+	if (usr_attr > 0)
+		H5Aclose(usr_attr);
+	if (attr_dtype > 0)
+		H5Tclose(attr_dtype);
+	if (attr_dspace > 0)
+		H5Sclose(attr_dspace);
+	return rc;
+}
+
+static int
+serialize_uint(hid_t file_id,
+	       struct daos_prop_entry *entry,
+	       const char *prop_str)
+{
+	int	rc = 0;
+	hid_t	status = 0;
+	hsize_t	attr_dims[1];
+	hid_t	attr_dtype = 0;
+	hid_t	attr_dspace = 0;
+	hid_t	usr_attr = 0;
+
+
+	attr_dims[0] = 1;
+	attr_dtype = H5Tcopy(H5T_NATIVE_UINT64);
+	if (attr_dtype < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	attr_dspace = H5Screate_simple(1, attr_dims, NULL);
+	if (attr_dspace < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	usr_attr = H5Acreate2(file_id, prop_str, attr_dtype,
+			      attr_dspace, H5P_DEFAULT, H5P_DEFAULT);
+	if (usr_attr < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	status = H5Awrite(usr_attr, attr_dtype, &entry->dpe_val);
+	if (status < 0) {
+		rc = -DER_IO;
+		D_ERROR("failed to write attr "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+out:
+	/* close hdf5 objects */
+	if (usr_attr > 0)
+		H5Aclose(usr_attr);
+	if (attr_dtype > 0)
+		H5Tclose(attr_dtype);
+	if (attr_dspace > 0)
+		H5Sclose(attr_dspace);
+	return rc;
+}
+
+int
+daos_cont_serialize_props(hid_t file_id, daos_prop_t *prop_query)
+{
+	int			rc = 0;
+	struct daos_prop_entry	*entry;
+
+	if (prop_query == NULL) {
+		rc = -DER_NOMEM;
+		D_GOTO(out, rc);
+	}
+
```

Comment:
question:
are we always guaranteed to get the properties, and all of them, in this order.
can this function get a subset of those properties to serialize, and in a different order?

Reply:
@mchaarawi yes the properties are always retrieved in the order that they are serialized. I guess there is an assumption here that all of the container properties are always serialized/written, but when deserializing and reading back into a new container not all of the properties are always used. This happens with container labels, which is always written/serialized, but only deserialized if the label no longer exists in the pool.

Reply:
right so if someone calls daos_cont_serialize_md() with let's say a prop with 2 entries then:
- some properties are not stored in the correct attributes.
- we can end up with a segfault?
my ultimate question here really is, wouldn't it be better to call serialize depending on prop type: dpe_type?
so something like:
if (prop->dpp_entries[0].dpe_type == DAOS_PROP_CO_LAYOUT_TYPE)
  serialize_uint(file, entry, "DAOS_PROP_CO_LAYOUT_TYPE");

you probably need the number of entries passed in too?
maybe im missing something here..